src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala

```diff
@@ -701,16 +701,18 @@ object CandidateGenerator extends ScorexLogging {
           if (!inputsNotSpent(tx, stateWithTxs) || doublespend(current, tx)) {
             //mark transaction as invalid if it tries to do double-spending or trying to spend outputs not present
             //do these checks before validating the scripts to save time
+            log.debug(s"Transaction ${tx.id} double-spending or spending non-existing inputs")
             loop(mempoolTxs.tail, acc, lastFeeTx, invalidTxs :+ tx.id)
           } else {
-            implicit val verifier: ErgoInterpreter = ErgoInterpreter(
+            val verifier: ErgoInterpreter = ErgoInterpreter(
```

Comment:
This verifier is created for each tx from mempoolTxs. 
In ErgoState.execTransactions it is created once for all transactions in the block.
It is impossible to do the same here (because we don't know which txs will be added to the block).
However, once the candidate is created (https://github.com/ergoplatform/ergo/blob/30c28eb66660eb0874cc10fd6ebadb00d4b62786/src/main/scala/org/ergoplatform/mining/CandidateGenerator.scala#L507) it can be verified again with ErgoState.execTransactions, i.e. with exactly the same procedure.

Reply:
Moved verifier outside the loop. 

ErgoState.execTransactions is called as part of local block processing. Calling it for candidate is possible, but could slow down miners a bit.

Reply:
> Moved verifier outside the loop.

I'm not sure it will work well. If the number of transactions in the pool is large, then all of the script will accumulate inside IRContext's internal data structures. 
So in this case resetContext is necessary before each transaction, same as in ErgoState.execTransactions

Reply:
resetContext is called for each transaction in validateStateful (which is in turn is called in statefulValidity which is called inside stateWithTxs.validateWithCost), pretty much as in ErgoState.execTransactions

Reply:
ah, ok, missed this.